src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Presentation/CustomCommitCompletion.cs

```diff
@@ -67,34 +70,24 @@ public override string Description
                 // to then decide to show the tooltip.  But once they decide to show the tooltip,
                 // they defer to us to get the contents for it asynchronously.  As such, we just want
                 // to give them something non-empty so they know to go get the async description.
-                if (this.CompletionItem.HasAsyncDescription)
-                {
-                    return "...";
-                }
-
-                return this.CompletionItem.GetDescriptionAsync(CancellationToken.None).WaitAndGetResult(CancellationToken.None).GetFullText();
+                return "...";
             }
         }
 
-        public string GetDescription_TestingOnly()
+        public async Task<CompletionDescription> GetDescriptionAsync(CancellationToken cancellationToken)
         {
-            return this.CompletionItem.GetDescriptionAsync(CancellationToken.None).WaitAndGetResult(CancellationToken.None).GetFullText();
+            var document = await GetDocumentAsync(cancellationToken).ConfigureAwait(false);
+            return await this.PresentationItem.GetDescriptionAsync(document, cancellationToken).ConfigureAwait(false);
         }
 
-        public TextRunProperties GetTextRunProperties(TextRunProperties defaultTextRunProperties)
+        private Task<Document> GetDocumentAsync(CancellationToken cancellationToken)
         {
-            var item = (CompletionItem as DescriptionModifyingCompletionItem)?.CompletionItem ?? CompletionItem;
-
-            return (item.CompletionProvider as ICustomCompletionItemFormatter)?.GetTextRunProperties(item, defaultTextRunProperties)
-                ?? defaultTextRunProperties;
+            return _completionPresenterSession.SubjectBuffer.CurrentSnapshot.AsText().GetDocumentWithFrozenPartialSemanticsAsync(cancellationToken);
```

Comment:
I'm a little concerned that this means the Document can be changing over time.  Is that desirable/intentional?